### ISSUE TITLE
update and elapsed won't follow UpdateFramerate if fixedTimestep is false

### ISSUE THREAD

**Issue by philipes** [HaxeFlixel/flixel#1885]
- **Flixel version:** 4.0.1
- **OpenFL version:** 3.6.1
- **Lime version:** 2.9.1
- **Affected targets:** All

________________________________

**Code snippet reproducing the issue**:

Main.hx: Note the 600 argument for UpdateFramerate.
```haxe
package;

import flixel.FlxGame;
import openfl.display.Sprite;

class Main extends Sprite
{
	public function new()
	{
		super();
		addChild(new FlxGame(640, 480, MenuState, 1, 600, 60, true));
	}
}

```

MenuState.hx
```haxe
package;

import flixel.FlxG;
import flixel.FlxSprite;
import flixel.FlxState;
import flixel.text.FlxText;
import flixel.ui.FlxButton;
import flixel.math.FlxMath;

class MenuState extends FlxState
{
	var elapseds: Array<Float>;
	
	override public function create():Void
	{
		super.create();
		
		FlxG.fixedTimestep = false;
		FlxG.debugger.visible = true;
		elapseds = [];
	}

	override public function update(elapsed: Float)
	{
		super.update(elapsed);
		
		/* 1 second when "fixedTimestep = true"
		 * 10 seconds when "fixedTimestep = false"
		 * */
		var size = 600;
		
		// Push the values
		if (elapseds.length < size) elapseds.push(elapsed);
		
		// Log the average
		else if (elapseds.length == size) 
		{
			var acc = 0.0;
			for (e in elapseds) acc += e;
			FlxG.log.add(1 / (acc / size));
			elapseds.push(0);
		}
	}
}

```

What seems off to me is [this line](https://github.com/HaxeFlixel/flixel/blob/dev/flixel/FlxGame.hx#L542) where step is called only once if fixedTimestep is false. Shouldn't it use a similar logic if fixedTimestep were true?

Also, I could not follow the exact logic of how elapse is calculated but it seems to use _elapsedMS which is updated only once per draw frame [here](https://github.com/HaxeFlixel/flixel/blob/dev/flixel/FlxGame.hx#L502).

________________________________

**Observed behavior:** Waits 10 seconds and logs a number around 60.
**Expected behavior:** Waits 1 second and logs a number around 600.

**Comment by MSGhero** [comment]
I believe it runs your game as fast as possible and uses whatever time step occurred between two frames.

You can't run 600 fps. If you're testing on flash, you can't go past 60. There might be a hard limit for other targets too.

**Comment by philipes** [comment]
Draw rate and update rate are two different things in HaxeFlixel.

Even if the draw rate is limited to 60, [here](https://github.com/HaxeFlixel/flixel/blob/dev/flixel/FlxGame.hx#L534) you can see that "step" will be called multiple times per draw frame to reach the desired update rate. This seems to happen only if FlxG.fixedTimestep is true, which I don't think is intended.

**Comment by sruloart** [comment]
This is not a bug: by design, a fixedTimeStep skips a couple of frames in order to keep the logic and the rendering synced, while a !fixedTimeStep (aka "variableTimeStep") slows/speeds up the logic by frame-time dependent interpolation.

**Comment by philipes** [comment]
Why? If I wanted them synced I'd pass the same value to UpdateFramerate and DrawFramerate, no?

**Comment by sruloart** [comment]
No. Different systems have different constraints. That means that over a period of time, a lag can be introduced, and this is why game-loops have ways to mitigate it. When you pass in the same value FPS what actually happens is that Flixel is fixing the lag for you, by either methods.

**Comment by philipes** [comment]
I'm not sure I understand.

By lag, do you mean an FPS drop? Isn't the point of FlxG.maxElapsed to avoid this problem?

**Comment by sruloart** [comment]
First question: Among other circumstances, yes. A good starting reading material: http://gameprogrammingpatterns.com/game-loop.html

Also: https://github.com/HaxeFlixel/flixel/blob/94fa8e4fb8eb666e7d9e6f4d2ab27ac72e97bab1/flixel/FlxGame.hx#L761-L775

Second question: No, .maxElapsed is meant to keep the interpolation from breaking when faced with a very high frame-time.

**Comment by philipes** [comment]
I have a fairly good understanding of how a game loop works. What confused me was your wording.

Your second link shows that elapsed uses the value of _elapsedMS which is constant within the same draw frame. Which is not a problem since step is called only once in the draw frame no matter the value UpdateFramerate.

--

Let me rephrase my problem.

In my understanding, UpdateFramerate dictates how many times the update function should be called any given second. I understand this might not happen because of lag, but HaxeFlixel should at least try to, which is not happening. This is the key issue I'm trying to point out.

Conceptually this is independent of DrawFramerate and fixedTimestep. I can see why this is not the case in the code, but I think it should.

**Comment by sruloart** [comment]
It can be implemented, I just don't think it should.

**Comment by philipes** [comment]
I have been using the following workaround for the past few days and I haven't encountered any problems. This could be some base for an implementation if there's any interest. 

I replaced [this line](https://github.com/HaxeFlixel/flixel/blob/dev/flixel/FlxGame.hx#L547), with:

```haxe
var stepsPerDrawFrame = Math.round(FlxG.updateFramerate / FlxG.drawFramerate);
if (stepsPerDrawFrame < 1) stepsPerDrawFrame = 1;

_elapsedMS /= stepsPerDrawFrame;
for (_ in 0...stepsPerDrawFrame)
	step();
```

Using the example above, "step" would be called 10 times with 1/10 of the elapsed time each.

I'm closing it for now. I'll submit a PR when I'm more familiar with the code base in the future.
